packages/create-pantheon-decoupled-kit/src/templates/gatsby-wp/src/components/paginator.tsx

```diff
@@ -14,125 +16,119 @@ import * as styles from './paginator.module.css'
  * currentItems is a subset of data, so any component that works for data will work here.
- * @param location holds flag variable to track breakpoint state
- * @param navRoute route of the page using the paginator
+ * @param props.location holds flag variable to track breakpoint state
+ * @param props.navRoute route of the page using the paginator
  * @returns Component with data rendered by the passed in Component and page links
  */
-const Paginator = ({
+const Paginator = <DataType,>({
 	data,
 	itemsPerPage,
 	breakpoints,
 	Component,
 	routing,
 	location,
-}) => {
-	// Track if breakpoint links have been opened
-	const openStart = location.state?.breakOpen
-		? breakpoints?.start + breakpoints?.add
-		: undefined
-
-	const [breakStart, setBreakStart] = useState(
-		openStart || breakpoints?.start || null,
-	)
-	const [breakpointsOpen, setBreakpointsOpen] = useState(
-		location.state?.breakOpen || false,
-	)
-	// This value will be the link to start with after separator
-	const breakEnd = breakpoints?.end || null
-	// how many links to add when the separator is clicked
-	const breakAdd = breakpoints?.add || null
-	const [currentPageQuery, setCurrentPageQuery] = useState(
-		routing && location.pathname.match(/\d+$/) !== null
-			? Number(location.pathname.match(/\d+$/)[0])
-			: 1,
-	)
-
-	// account for extra forward slash added to url on hard refresh
-	const currentPath = location.pathname.endsWith('/')
-		? location.pathname.replace(/.$/, '')
-		: location.pathname
-	const navRoute = routing ? currentPath.replace(/\/\d+$/, '') : currentPath
-
-	const [offset, setOffset] = useState((currentPageQuery - 1) * itemsPerPage)
-	const [currentItems, setCurrentItems] = useState([])
-	const [totalItems] = useState(data.length)
-	const [totalPages] = useState(Math.ceil(data.length / itemsPerPage))
-
-	// since we fetch ALL items, filtering the current items client side
-	// further reduces API calls to the server
+}: PaginatorProps<DataType[]>) => {
+	const {
+		breakStart,
+		setBreakStart,
+		breakpointsOpen,
+		setBreakpointsOpen,
+		breakEnd,
+		breakAdd,
+		currentPageQuery,
+		setCurrentPageQuery,
+		navRoute,
+		offset,
+		setOffset,
+		currentItems,
+		setCurrentItems,
+		totalItems,
+		totalPages,
+	} = usePagination({ data, breakpoints, itemsPerPage, routing, location });
 	useEffect(() => {
 		setCurrentItems(
-			data.slice(
+			data?.slice(
 				(currentPageQuery - 1) * itemsPerPage,
 				itemsPerPage * currentPageQuery,
 			),
-		)
-
-		// set the new offset
+		);
+		// If we're on page 1 or greater, set the new offset to
+		// ((currentPageQuery - 1) * itemsPerPage) % totalItems
 		const newOffset =
-			((currentPageQuery > 0 && currentPageQuery - 1) * itemsPerPage) %
-			totalItems
-		setOffset(Number(newOffset))
-	}, [data, offset, itemsPerPage, breakStart, currentPageQuery, totalItems])
+			(Math.max(currentPageQuery - 1, 0) * itemsPerPage) % totalItems;
```

Comment:
I think something may be off with this new page equation. I am experiencing this regression when trying to access the last page of the pagination example. When first accessing the 10th page, the correct paginated data did show but the URL was still on `/9`. After loading for a bit the URL jumped to `/10` and threw this error. 

https://github.com/pantheon-systems/decoupled-kit-js/assets/39682014/f6371e85-056f-45c3-a332-7bb87058929a

I generated a new `gatsby-wp` starter from canary and was not able to recreate this there.